### PR TITLE
fix cursor movement

### PR DESCRIPTION
**Pull request by anuejn**
- ♻️ make player api declarative
- 🧪️ add test for editor/goLeft
- 🐛️ fix cursor not being re-positioned
- 🐛️ fix itertools.dropwhile
- 🐛️ fix cursor left / right movement
- 🧪️ add tests for cursor left / right movement

### CODE REVIEW

**Comment by anuejn** [comment]
this should close #217

**Comment by anuejn** [comment]
this depends on #221

File: app/src/util/itertools.ts [hunk 1]
```diff
@@ -101,13 +109,13 @@ function* filter<T>(predicate: (x: T) => boolean, input: Generator<T>): Generato
 }
 
 function* dropwhile<T>(predicate: (x: T) => boolean, input: Generator<T>): Generator<T> {
+  let dropping = true;
   for (const item of input) {
-    if (!predicate(item)) {
+    dropping = dropping && predicate(item);
+    if (!dropping) {
```
Comment (pajowu) [inline app/src/util/itertools.ts:115]: Nice catch, I did not know that generators are not reusable (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of#do_not_reuse_generators). Please also update `takewhile`

File: app/src/state/editor/play.ts [hunk 2]
```diff
@@ -32,12 +32,21 @@ export const togglePlaying = createActionWithReducer<EditorState>(
...
-  )[0];
-  assertSome(item);
-  state.currentTimeUserSet = item.absoluteStart;
+  );
+  const firstItem = items[0];
+  const secondItem = items[items.length - 1];
+  assertSome(firstItem);
+  assertSome(secondItem);
+
+  // if we are at the beginning of a paragraph, we should put the cursor at the end of the paragraph before
```
Comment (pajowu) [inline app/src/state/editor/play.ts:44]:
```suggestion
  // if we are at the beginning of a paragraph, we should put the cursor at the end of the previous paragraph
```

(I had to read this sentence twice, I personally think this makes it clearer)

**Review by pajowu — CHANGES_REQUESTED** [review]

**Review by pajowu — APPROVED** [review]

**Pull request merged**